src/main/java/de/tum/in/www1/artemis/service/user/UserService.java

```diff
@@ -662,7 +663,16 @@ public Optional<User> findUserAndAddToCourse(String registrationNumber, String c
                 return optionalStudent;
             }
 
-            log.warn("User with registration number '{}' and login '{}' not found in Artemis user database nor found in (TUM) LDAP", registrationNumber, login);
+            // 4) if we still cannot find the user, we try again using the email
```

Comment:
💡Idea: You could greatly reduce the code duplication in this method, by writing:
```
            var optionalStudent = 
                userRepository.findUserWithGroupsAndAuthoritiesByRegistrationNumber(registrationNumber)
                    .or(() -> this.createUserFromLdap(registrationNumber))
                    .or(() -> userRepository.findUserWithGroupsAndAuthoritiesByLogin(login))
                    .or(() ->  userRepository.findUserWithGroupsAndAuthoritiesByEmail(email));
            
            
            if (optionalStudent.isPresent()) {
                var student = optionalStudent.get();
                // the newly created user needs to get the rights to access the course
                this.addUserToGroup(student, courseGroupName, courseGroupRole);
                return optionalStudent;
            }
```

Reply:
Very good refactoring suggestion, thanks 😁
Added it with https://github.com/ls1intum/Artemis/pull/5722/commits/5e90ac4b0499884f48d7b066899ab045e2aa2760